File: Grind75_1/week5/MergeIntervals.py
"""
https://leetcode.com/problems/merge-intervals/description/
"""

class Solution:

    def intervalMerger(self, arr: List[List[int]], iter1: int, currStart: int, currEnd: int) -> tuple():

        pop = False
        if (arr[iter1][0] >= currStart and arr[iter1][0] <= currEnd):
            # beginning of the interval is within the current interval
            pop = True
            
            currEnd = max(currEnd, arr[iter1][1])

        elif (arr[iter1][1] >= currStart and arr[iter1][1] <= currEnd):
            # end of the interval is within the current interval
            pop = True
            
            currStart = min(currStart, arr[iter1][0])

        elif (arr[iter1][0] < currStart and arr[iter1][1] > currEnd):
            # interval totally encapsulates current interval
            pop = True

            currStart = arr[iter1][0]

            currEnd = arr[iter1][1]

        return (currStart, currEnd, pop)


    # time exceeded, need to try again later
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        
        retList = []

        currStart = None
        currEnd = None

        idx = 0
        lenRet = len(retList)
        lenArr = len(intervals)

        while (idx < lenArr):
            # loop with current element
            curr = intervals.pop(idx)
            currStart = curr[0]
            currEnd = curr[1]

            lenArr = len(intervals)

            iter1 = 0
            # check all elements in case there is no order in the original list
            while (iter1 < lenArr):

                if (iter1 == idx):
                    iter1 = iter1 + 1
                    continue
                    
                resTuple = self.intervalMerger(intervals, iter1, currStart, currEnd)
                
                currStart = resTuple[0]
                currEnd = resTuple[1]

                if (resTuple[2] == True):
                    p = intervals.pop(iter1)
                    lenArr = len(intervals)
                else:
                    iter1 = iter1 + 1

            # also check retList to see if can merge
            iter1 = 0
            lenRet = len(retList)

            while (iter1 < lenRet):
                resTuple = self.intervalMerger(retList, iter1, currStart, currEnd)

                currStart = resTuple[0]
                currEnd = resTuple[1]

                if (resTuple[2] == True):
                    p = retList.pop(iter1)
                    lenRet = len(retList)
                else:
                    iter1 = iter1 + 1

            
            # finally append to retList
            retList.append([currStart, currEnd])


        return retList

        

File: neetcode/blind75_JS/Intervals/merge-intervals.js
// https://neetcode.io/problems/merge-intervals

class Solution {
    /**
     * @param {number[][]} intervals
     * @return {number[][]}
     */
    merge(intervals) {
        if (intervals.length < 2) {
            return intervals
        }
        // console.log(intervals)
        intervals.sort((a, b) => {return a[0] - b[0]})

        const res = []

        while (intervals.length) {
            let i = 0
            let j = 1
            while (j < intervals.length && intervals[i][1] >= intervals[j][0]) {
                intervals[i][0] = Math.min(intervals[i][0], intervals[j][0])
                intervals[i][1] = Math.max(intervals[i][1], intervals[j][1])
                j += 1;
            }
            res.push(intervals[i])

            intervals = intervals.slice(j)
        }

        return res
    }
}
